### PR TITLE
ARROW-3722: [C++] Allow specifying types of CSV columns

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Comment by alendit** [comment]
Looks good to me.

**Comment by pitrou** [comment]
I'm not sure what happened with AppVeyor here. I'm trying to get a clean CI build on my AppVeyor account.

File: python/pyarrow/tests/test_csv.py [hunk 1]
```diff
@@ -200,6 +208,30 @@ def test_simple_nulls(self):
             'e': [b"3", b"nan", b"\xff"],
             }
 
+    def test_column_types(self):
+        # Ask for specific column types in ConvertOptions
+        opts = ConvertOptions(column_types={'b': pa.float32(),
+                                            'c': pa.string(),
+                                            'd': pa.null()})
```
Comment (dhirschfeld) [inline python/pyarrow/tests/test_csv.py]: If I already have a `pa.Schema` object which defines the column types can I pass that directly to `pa.csv.read_csv`?

Comment (pitrou) [reply python/pyarrow/tests/test_csv.py]: No. Perhaps we want to allow passing a schema as `column_types`? @wesm

Comment (dhirschfeld) [reply python/pyarrow/tests/test_csv.py]: From a user api point-of-view I think that would be easier than requiring a mapping:
```python
column_types = {field.name: field.type for field in schema}
```
i.e. the schema should be the one way to describe arrow Tables.

A mapping api is made a bit more cumbersome because there's no trivial way to conver to to a schema (I think that would a useful function/method to have):
```python
schema = pa.schema(list(starmap(pa.field, column_types.items())))
```

Comment (wesm) [reply python/pyarrow/tests/test_csv.py]: I think it would be good to also accept a Schema, it's another way to indicate the desired column types

Comment (wesm) [reply python/pyarrow/tests/test_csv.py]: Some CSV files have hundreds of columns. Take a look at the API for this in the R readr package https://cran.r-project.org/web/packages/readr/vignettes/readr.html

Comment (dhirschfeld) [reply python/pyarrow/tests/test_csv.py]: > *We will require configurability for some types beyond just a data type eventually (e.g. categorical with known categories)*

I'm not sure I follow - isn't a categorical type an actual data type which is defined by its categories?

Comment (wesm) [reply python/pyarrow/tests/test_csv.py]: Sorry, bad example. Let me try again: dates with a particular format (like "%d-%m-%y")

Comment (dhirschfeld) [reply python/pyarrow/tests/test_csv.py]: Yeah, this (applying a known schema) and date handling are the most important features for me! :)

**Comment by wesm** [comment]
Looking

**Review by wesm — COMMENTED** [review]
A minor thing around accepting type aliases (this would improve usability), but otherwise LGTM

File: python/pyarrow/_csv.pyx [hunk 2]
```diff
@@ -266,6 +277,36 @@ cdef class ConvertOptions:
...
+            value = value.items()
+
+        self.options.column_types.clear()
+        for item in value:
+            if isinstance(item, Field):
+                k = item.name
+                v = item.type
+            else:
+                k, v = item
+            typ = pyarrow_unwrap_data_type(v)
```
[the remark this answers is not in the document]
Comment (pitrou) [reply python/pyarrow/_csv.pyx]: Cool, I didn't know about that function :-)

**Comment by pitrou** [comment]
Gandiva failure on Travis-CI is unrelated.
